pkg/dfns-api-client/auth_transport.go

```diff
@@ -0,0 +1,286 @@
+package dfns_api_client
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+
+	"github/dfns/dfns-sdk-go/pkg/credentials"
+)
+
+const (
+	UserActionHeader    = "x-dfns-useraction"
+	appIDHeader         = "x-dfns-appid"
+	nonceHeader         = "x-dfns-nonce"
+	appSecretHeader     = "x-dfns-appsecret" //nolint:gosec
+	authorizationHeader = "authorization"
+	contentTypeHeader   = "Content-Type"
+)
+
+type authTransport struct {
+	*DfnsBaseApiOptions
+}
+
+func (auth *authTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	auth.setHeaders(req)
+
+	methodsWithAuth := map[string]struct{}{
+		"POST":   {},
+		"PUT":    {},
+		"DELETE": {},
+	}
+
+	if _, ok := methodsWithAuth[req.Method]; ok {
+		performUserAction, err := shouldPerformUserAction(req)
+		if err != nil {
+			return nil, err
+		}
+
+		if performUserAction {
+			if err := auth.performUserActionRequest(req); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	return executeRequest(req)
+}
+
+func shouldPerformUserAction(req *http.Request) (bool, error) {
+	authHeaderStr := req.Header.Get(UserActionHeader)
+	if authHeaderStr == "" {
+		return true, nil
+	}
+	defer req.Header.Del(UserActionHeader)
+
+	authHeader, err := strconv.ParseBool(authHeaderStr)
+	if err != nil {
+		return false, fmt.Errorf("error parsing header value as boolean: %w", err)
+	}
+
+	return authHeader, nil
+}
+
+// PerformSimpleRequest performs a simple HTTP request.
+func (auth *authTransport) performSimpleRequest(req *http.Request) (*http.Response, error) {
+	auth.setHeaders(req)
+
+	return executeRequest(req)
+}
+
+// PerformUserActionRequest performs a user action HTTP request.
+func (auth *authTransport) performUserActionRequest(req *http.Request) error {
+	auth.setHeaders(req)
+
+	return auth.executeUserActionRequest(req)
+}
+
+// setHeaders sets the request headers.
+func (auth *authTransport) setHeaders(req *http.Request) {
+	req.Header.Set(appIDHeader, auth.AppId)
+	req.Header.Set(nonceHeader, generateNonce())
+
+	if auth.AppSecret != nil {
+		req.Header.Set(appSecretHeader, *auth.AppSecret)
+	}
+
+	if auth.AuthToken != nil {
+		req.Header.Set(authorizationHeader, "Bearer "+*auth.AuthToken)
+	}
+
+	req.Header.Set(contentTypeHeader, "application/json")
+}
+
+func executeRequest(req *http.Request) (*http.Response, error) {
+	response, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	// Fill Request otherwise it's nil
+	response.Request = req
+
+	if err := handleResponseError(response); err != nil {
+		return nil, err
+	}
+
+	return response, nil
+}
+
+// handleResponseError handles errors in the HTTP response.
+func handleResponseError(response *http.Response) error {
+	if response.StatusCode < 400 || response.StatusCode == PolicyPendingErrorCode {
```

Comment:
I think it should be

```suggestion
	if response.StatusCode < 400 && response.StatusCode != PolicyPendingErrorCode {
```

Reply:
arf.... good catch